### ISSUE TITLE
Re-licensing libzmq

### ISSUE THREAD

**Issue by rlenferink** [zeromq/libzmq#2376]
Hello libzmq contributors,

I'm starting this issue because of the current libzmq license. At the moment, libzmq is licensed as LGPL with static linking exception and this makes it a complex "one of a kind" license.

Almost a year ago the ZeroMQ project team started to collect license grants to re-license libzmq to a clearer license, namely the MPLv2. Collecting these license grants stopped for a while, because Pieter Hintjens, the former BDFL (Benevolent Dictator for Life) behind ZeroMQ, got sick and passed away.

We want to continue Pieter's work with collecting these contributor license grants. We want to ask you to send us a license grant to permit the libzmq to relicense to a different license.  At the moment, libzmq is licensed as LGPL 3 with a static linking exception (http://zeromq.org/area:licensing).  This makes it hard to make use of libzmq in projects with some other Open Source licenses (e.g. projects using the Apache License).  The existing LGPL 3 with exceptions is also a non-standard license that some corporate lawyers are unwilling to accept for use in their companies.

For relicensing the libzmq project, we need to collect a license grant from each individual contributor who wrote a major piece of code in the development process of libzmq.

We are requesting contributors to send in a license grant to relicense libzmq to one of the options listed below:
1. relicense to the MPLv2 license specifically;
1. relicense to the MPLv2 or any "share alike" Open Source Initiative (OSI) approved license chosen by the current ZeroMQ BDFL;
1. relicense to the MPLv2 or any Open Source Initiative (OSI) approved license chosen by the current ZeroMQ BDFL **(preferred selection)**

If a contribution was made during working hours for an employer then we will also need a license grant (or "quit claim") from the employer.

We would like to follow a certain template as license grant to ensure all the necessary information is included.  There are three templates, one for each option.  Please choose the template that best represents your wishes.

The templates can be found at: https://github.com/zeromq/libzmq/tree/master/RELICENSE/templates

To include the license grant, please:
- fork the original libzmq repository
- create a new file in the RELICENSE directory, named after your name and/or your employers name, with the license grant as content, using the template chosen
- commit this file to the repository, using the same email address as your earlier contributions
- create a pull request on libzmq to have your license grant merged into the repository.

See the existing files in:
https://github.com/zeromq/libzmq/tree/master/RELICENSE

or this specific example: https://github.com/zeromq/libzmq/blob/master/RELICENSE/rlenferink.md

for a guide for how to fill out the template.

---

If there are any questions, don't hesitate to ask them below.

Roy

**Comment by bluca** [comment]
Thanks for kickstarting the process again!

**Comment by vielmetti** [comment]
I contributed a one-line update to some documentation, is it necessary to go through an entire license grant to handle that?

**Comment by rlenferink** [comment]
@vielmetti this is the specific commit you contributed: https://github.com/zeromq/libzmq/commit/92b1b2b18184b63bdd1586df5b46c586d54f200f

Since it is only documentation and not a major addition to libzmq it is not required for you to send in a license grant.

**Comment by junovitch** [comment]
Hi, I believe my submissions were two fixes to fix compiling the FreeBSD port of zeromq. One for GCC not liking a missing newline (https://github.com/zeromq/zeromq4-x/commit/2ac32f8c49188e9d251081d9d6e4e00499fce31e) and one for fixing compiling with the documentation option by just shifting code (https://github.com/zeromq/zeromq4-1/commit/c75cc9ec5b8886b291e3ad8b8a05b6d255822c8f). I personally don't consider either significant enough of a contribution.

Good on you folks. Good luck!

**Comment by Asmod4n** [comment]
I believe i don't have to sign it, or? just added a option to zmq_has afaik, but how can i find out my contributions?

**Comment by bluca** [comment]
https://github.com/zeromq/libzmq/commits?author=Asmod4n

It's a few lines, so if you don't mind it would be better to. Thanks!

**Comment by Asmod4n** [comment]
Could someone elaborate what the 2nd or 3rd options could mean? I currently like the MPLv2 the most, because its compatible to pretty much every other license?

**Comment by bluca** [comment]
I like MPL2 as well and it's 99/99% what is going to end up being, but given we are going through this ordeal it's worth, if the contributor agrees, to keep the options open.
It's up to you what you prefer for your grant.

**Comment by lresende** [comment]
Where are we with the relicensing process? Any ETA?

**Comment by bluca** [comment]
No ETA, there are still many old contributors missing.

**Comment by bluca** [comment]
@sigiesec let's try to be more disciplined on this - GitHub actually helps and tells if someone is a "first time contributor" so when we see that, let's ask for the grant before merging

**Comment by sigiesec** [comment]
@bluca Ok, I will try too :)

**Comment by rlenferink** [comment]
It's been some time ago since I've been working on this topic as well. At the time of sending out the first request to submit a grant, I've kept track of the responses but it should be updated to include new contributors from the past couple of months as well. I'm still busy this week but will try to restart this process somewhere next week.

As @bluca mentioned it indeed would be good to ask first time contributors to submit a grant before merging ;)

**Comment by bluca** [comment]
@rlenferink hi - I am looking at this again. Did you have a script or set of scripts to compile the checklist? I'd be happy to run them myself

**Comment by bluca** [comment]
@hurtonm hi Martin - we were looking at the history of libzmq commits for the relicensing effort and we _think_ your commits were copyright of iMatix - could you please confirm that? Thanks!

**Comment by hurtonm** [comment]
Hi Luca,  you are right, I confirm that.

Best regards
Martin Hurtoň

**Comment by bluca** [comment]
Thank you!

**Comment by rlenferink** [comment]
hi @bluca thanks for taking your time as well (and good to remind me of this).

I put the instructions in [libzmq-relicense](https://github.com/rlenferink/libzmq-relicense)

**Comment by bluca** [comment]
256 grants collected, 14 remaining to collect as documented on: https://github.com/rlenferink/libzmq-relicense/blob/master/checklist.md

**Comment by Dysl3xik** [comment]
I see you guys are working on updating the license terms for this project. On the main page you list that the liccense is LGPL currently but the github license link seems to point to the current license being GPLv3 instead.

See https://github.com/zeromq/libzmq/blob/master/COPYING
and https://github.com/zeromq/libzmq/blob/master/README.md 

Can this be clarified to remove any doubt while you work in the re-license?

**Comment by bluca** [comment]
it's common practice with LGPL to include both texts - https://github.com/zeromq/libzmq/blob/master/COPYING.LESSER

**Comment by Dysl3xik** [comment]
> it's common practice with LGPL to include both texts - https://github.com/zeromq/libzmq/blob/master/COPYING.LESSER

Thats fair, but someone who clicks on the license link on the project main page would be taken straight to the GPL instead of the LGPL - this could scare them away from the project

![image](https://user-images.githubusercontent.com/21955286/102694940-475b7800-41f2-11eb-8eb4-d818c8b07235.png)

**Comment by bluca** [comment]
I cannot affect what Github parses and decides to link on its own. I don't think you can change what that points to. The documentation, license texts and copyright blurbs in all source files are all crystal clear.

https://github.com/zeromq/libzmq#license

**Comment by johanjeppsson** [comment]
Currently, we are not able to use zmq in our corporate application due to the LGPL licence, so this is a great initiative. 

It seems like progress has slowed down quite a bit due to old contributors being hard to reach. What's the plan if the final grants cannot be collected?

**Comment by bluca** [comment]
There's not really a plan, as the last ~5 contributors have code too intertwined to rewrite and too old to be able to tell whether a relicense is required or it can be skipped without paying a lawyer to do a thorough analysis. BTW why is exactly LGPL a problem for your usage? It's used in loads of corporate applications

**Comment by johanjeppsson** [comment]
>  BTW why is exactly LGPL a problem for your usage? It's used in loads of corporate applications

 I'm not sure as to why, unfortunately. We have a list of approved licences, and  LGPLv3 is not one of those.

**Comment by bluca** [comment]
you'll want to double check with your lawyers, as this is LGPL + exceptions for the static linking and tivoization clauses, so for all intents and purposes there is no pratical difference from the MPL2 which will be the new license eventually

**Comment by johanjeppsson** [comment]
I actually did ask specifically about the exception from the LGPL license when we were looking into this, and their answer was still no. 

I realize that the new license makes little difference from a legal standpoint, but I guess that it is easier for legal teams to deal with standard licences; which I guess is also one of the reasons behind this relicensing endeavor.

**Comment by bluca** [comment]
Yes indeed that's one of the main reasons, although I think this is the first time I hear of it being actually an issue. I wish I had better news for you, but I regularly contact the individuals every few months and have been doing so for a few years, with no luck so far.

**Comment by brugr9** [comment]
A project of mine has just been rejected by a major software vendor because the (customized) lgpl discouraged - despite the statically linked libzmq ... :(
What about the venture to move to MPLv2 - any news?

**Comment by bluca** [comment]
I periodically contact the last 6 people that we need a relicense grant from, unfortunately with no success. You can see the list here:

https://github.com/rlenferink/libzmq-relicense/blob/master/checklist.md#grants-to-be-submitted

Unfortunately there's really nothing one can do until either those 6 grants are received, or someone else clean-room reimplements all the relevant changes by those contributors.

**Comment by brugr9** [comment]
Thank you very much for your quick reply, which I appreciate very much.
Ok, but that looks good in principle, doesn't it?

Eventually DM via another channel could help?
- Laurent Alebarde 
  - https://www.linkedin.com/in/laurent-alebarde-93303b22/
- Thijs Terlouw 
  - https://www.linkedin.com/in/thijsterlouw/
  - https://twitter.com/startinchina
 - Patrick Trantham
   - https://www.linkedin.com/in/patrick-trantham-b7b0a262/
- Hitster GTD
  - https://twitter.com/hitstergtd
- Guido Goldstein
  - https://www.linkedin.com/in/guido-goldstein-31aa7b1a6/

**Comment by brugr9** [comment]
@bluca It is noticeable that on stackoverflow Hitster GTD was "Last seen more than 5 years ago" (2016) - last action was in April 2016 (cp.  https://stackoverflow.com/users/6019885/hitster-gtd?tab=activity ). Is it possible that 'Hitster GTD' was a pseudonym of Pieter Hintjens?

**Comment by soundart** [comment]
I think I explained it the last time using email already: I am not happy with this activity. 

That being said, these three contributions are trivial, easy to replace and in no way interwoven in a way that they cannot be  replaced easily.

Please revert these and reimplement them and you are done.
   Merge pull request #1018 from soundart/master
   Merge pull request #936 from soundart/master
   Merge pull request #933 from soundart/master

This one integrates code not from me, but from djb.
    Merge pull request #997 from soundart/master

**Comment by brugr9** [comment]
> I think I explained it the last time using email already: I am not happy with this activity.
> [...]

Thank you for your answer here directly in the issue - that shows respect.

**Comment by somdoron** [comment]
@brugr9 Hitster is not pseudonym for Pieter, I met him once. Anyway,. I think his PRs are insignificant, mostly about cleaning code, docs or formatting.

**Comment by ricnewton** [comment]
@brugr9 FYI, I'm not @guidog, I guess just another Taz fan :).

**Comment by bluca** [comment]
> I think I explained it the last time using email already: I am not happy with this activity.

My apologies, I have contacted literally hundreds of developers so losing track of unfinished conversations can happen, and it's what happened here. As mentioned in my old email, the effort is not about commercial adoption, but to have the same effective terms and conditions without a custom license (which is bad, not tested in court, and complicated to explain). Our customized LGPL3 is for all intents and purposes the same as the vanilla Mozilla Public License v2 - ie, share-alike for our code, can be linked statically or dynamically wherever.

> That being said, these three contributions are trivial, easy to replace and in no way interwoven in a way that they cannot be replaced easily.
> 
> Please revert these and reimplement them and you are done. Merge pull request #1018 from soundart/master Merge pull request #936 from soundart/master Merge pull request #933 from soundart/master
> 
> This one integrates code not from me, but from djb. Merge pull request #997 from soundart/master

I'm afraid it's not that simple. Given the explicit NACK, the only safe way would be to completely remove tweetnacl, and also reintroduce those bugs you fixed. And then wait for somebody to contribute bug fixes while being able to prove they never looked at the previous fixes (which excludes most if not all existing maintainers and contributors), and same for the tweetnacl integration. Otherwise the legality of the license change to MPL2 would be forever in doubt - which in most environments it might as well mean it is invalid, as the point is to remove uncertainty, not add to it.

**Comment by bluca** [comment]
> @brugr9 Hitster is not pseudonym for Pieter, I met him once. Anyway,. I think his PRs are insignificant, mostly about cleaning code, docs or formatting.

Yeah it's another person, I had email back and forth, will send another. Unfortunately it's not that simple in that case either - there will be always the doubt of whether any of the current CMake infrastructure is partially derived work including his changes. The only safe way would be to completely remove CMake, and for someone who's never looked at our implementation to provide integration done from scratch.

Again the goal here is to remove any room whatsoever for any possible doubt, and for the situation to be crystal clear. Otherwise it won't help anybody.

**Comment by brugr9** [comment]
@somdoron Nice to meet you - how do you do? 
@somdoron @bluca Thank you both for clarifying the pseudonym question.

And: 
> Unfortunately it's not that simple in that case either - there will be always the doubt of whether any of the current CMake infrastructure is partially derived work including his changes. The only safe way would be to completely remove CMake

Ok, how about splitting the libzmq repo by re-moving the current libzmq CMake infrastructure into a separate new project?
would this further relicensing if @hitstergtd does not report?

**Comment by bluca** [comment]
> > Unfortunately it's not that simple in that case either - there will be always the doubt of whether any of the current CMake infrastructure is partially derived work including his changes. The only safe way would be to completely remove CMake
> 
> Ok, how about splitting the libzmq repo by re-moving the current libzmq CMake infrastructure into a separate new project? would this help re-licensing if @hitstergtd does not report?

That would not make any difference

**Comment by bluca** [comment]
> > > Unfortunately it's not that simple in that case either - there will be always the doubt of whether any of the current CMake infrastructure is partially derived work including his changes. The only safe way would be to completely remove CMake
> > 
> > 
> > Ok, how about splitting the libzmq repo by re-moving the current libzmq CMake infrastructure into a separate new project? would this help re-licensing if @hitstergtd does not report?
> 
> That would not make any difference

Got permission to relicense to MPLv2 from him today, so not a problem anymore.

**Comment by bluca** [comment]
> * Wojciech Kula aka Cziken
>   * [x]  contacted via https://www.linkedin.com/in/wojciech-kula/ 06/01/2022
>   * [x]  has responded 07/01/2022

@brugr9 what do you mean here by "has responded"?

**Comment by brugr9** [comment]
> what do you mean here by "has responded"?

@bluca Wojciech Kula has answered on LinkedIn: _Friday 07/01/2022, 07:51 &mdash; Sure. Will do after coming back from holiday. Sorry for trouble_

**Comment by brugr9** [comment]
@bluca I wrote to him again and asked for time details:
> Dear Mr Kula, 
> Sorry again for the disturbance. May I kindly ask when you will be back from vacation resp. when you want to do the edit on github?
> So that we can plan the further procedure for re-licensing ...
> Thank you in advance for your understanding.
> Kindest Regards,
> Roland Bruggmann

I'm glad to see that things are moving forward.

**Comment by brugr9** [comment]
@bluca Mr Kula has kindly already replied:
> I will be in front of my computer on Tuesday morning and it will be the first thing I will do, trust me. I'm sorry for all the problems you have faced because of my negligence 

And I thanked Him for that:
> Thank you for responding so quickly - that is very kind of you. So we are looking forward to next week. Have good weekend!
> Best regards, Roland Bruggmann

**Comment by brugr9** [comment]
@bluca @somdoron As one can remark @soundart contributed for curve_server and curve_client exclusively &mdash; a few small though important edits. Can these curve features of libzmq be replaced by libcurve?

**Comment by bluca** [comment]
Not really

**Comment by brugr9** [comment]
@soundart My dear, "What's Up" and "Bam-ba-Lam" &mdash; you wrote:

> I am not happy with this activity.

Please allow me &mdash; may I still ask nicely: What are you missing in the MPLv2 to be happy, respectively, what are you missing to be happy with the MPLv2? It can be guessed that the MPLv2 is in the sense of the founder of this software&mdash;since he has used them himself. It would be really helpful if you would cooperate.

**Comment by brugr9** [comment]
@bluca Did @thijsterlouw ACK for MPLv2? I ask because #145 was closed but @thijsterlouw is still under [checklist#grant-necessary](https://github.com/rlenferink/libzmq-relicense/blob/54c53a08440419bc9537b0668e91684943952443/checklist.md#grant-necessary).

**Comment by bluca** [comment]
those patches were applied manually

**Comment by brugr9** [comment]
@soundart We still don't know what is preventing you from agreeing to the MPLv2. Martin Sustrik’s aka @sustrik &mdash;who submitted a lot to libzmq&mdash;has initiated nanomsg precisely for licensing reasons. He wrote (cp. [Why Communication Infrastructure Should Use Permissive Licenses](https://250bpm.com/blog:15/)):

_Consider following use cases that would not be possible with GPL'd codebase:_

- _Putting the technology into BSD kernel(s)._
- _Incorporating the technology into Windows OS._
- _Providing the infrastructure on iPhones and iPads._
- _Even using ZeroMQ (LGPLv3) code in Linux (GPLv2) in-kernel implementation would be a problem._
- _Supporting the infrastructure inside networking hardware._
- _Plugging the proprietary products into the infrastructure_

Is there a reason why libzmq has to degenerate into a Museeums piece?

**Comment by brugr9** [comment]
@bluca As seen on @lalebarde account he was active on github this year. Did @lalebarde already respond concerning the re-licensing libzmq to MPLv2?

**Comment by bluca** [comment]
No answer so far

**Comment by nightlark** [comment]
Instead of continuing to ask for relicensing statements from new PRs, it might be worth exploring the [LLVM relicensing](https://foundation.llvm.org/docs/relicensing/) approach where all contributions after `X` date (which for LLVM was `January 19, 2019 after the LLVM 8.0 release was branched`). Then wait until either the remaining [relicensing long tail](https://foundation.llvm.org/docs/relicensing_long_tail/) responses to trickle in, or rewritten code takes the place of the old contributions.

**Comment by bluca** [comment]
I have done some work to remove contributions from the authors who did not grant permissions:

https://github.com/zeromq/libzmq/pull/4554

The two removed APIs can be made non-stubs again if somebody is able to provide a clean-room reimplementation (ie: without looking at the old code that was removed). I can review such implementations, but for obvious reasons I cannot provide them, as my knowledge is 'tainted'.
With these changes, the relicense can be finalized.

**Comment by bluca** [comment]
> _Consider following use cases that would not be possible with GPL'd codebase:_
> 
> * _Putting the technology into BSD kernel(s)._
> * _Incorporating the technology into Windows OS._
> * _Providing the infrastructure on iPhones and iPads._
> * _Even using ZeroMQ (LGPLv3) code in Linux (GPLv2) in-kernel implementation would be a problem._
> * _Supporting the infrastructure inside networking hardware._
> * _Plugging the proprietary products into the infrastructure_

Note that none of this is true. The new license will be MPL2, which is very much intentionally still share-alike, and in layman's terms, for all intents and purposes, provides the same terms as the current LGPL3 + custom exceptions. The main purpose is to use a 'standard' license (no custom exceptions) and harmonize with the rest of the zeromq projects which are already on MPL2.

e.g.: the current license already allows using libzmq in proprietary products or networking hardware, and it already routinely happens.

**Comment by nightlark** [comment]
> I have done some work to remove contributions from the authors who did not grant permissions:
> 
> #4554
> 
> The two removed APIs can be made non-stubs again if somebody is able to provide a clean-room reimplementation (ie: without looking at the old code that was removed). I can review such implementations, but for obvious reasons I cannot provide them, as my knowledge is 'tainted'. With these changes, the relicense can be finalized.

Should clean-room reimplementations of the removed features be backwards compatible in terms of how they are used? Eg code setting the CMake option for using tweetnacl, or using the reimplemented functions should continue to work without the user having to make change to their code to use the new MPL implementations?

**Comment by bluca** [comment]
Yes - note that I am referring to zmq_proxy_steerable() and ZMQ_RECONNECT_IVL_MAX. I don't see any point in adding back tweetnacl, it just causes problems and it's unsafe to use anyway. libsodium is available everywhere. I don't think anybody should spend any time on tweetnacl, but obviously I can't stop anyone from doing so.

**Comment by bluca** [comment]
@lalebarde I see that you were active in the past week on Github, so one last chance to please give us permission to relicense your libzmq contributions to MPL-2.0. Your contributions are otherwise going to be deleted, and you will be removed from the zeromq Github organization.

**Comment by bluca** [comment]
Relicensing has now been completed.

**Comment by bluca** [comment]
@rlenferink could you please transfer ownership of the https://github.com/rlenferink/libzmq-relicense repository to me? I will then transfer it to the org here, and archive it

**Comment by shripchenko** [comment]
Prob im - the author of ZMQ_RECONNECT_IVL_MAX
should i do anything ?
i can,

**Comment by bluca** [comment]
@shripchenko assuming you have not sent a relicense grant, you could send one just to clarify any doubt, simply copy the following text in a new comment filling with your data, I'll do the rest:

```
# Permission to Relicense under MPLv2

This is a statement by {{ name of company / name of individual }}
that grants permission to relicense its copyrights in the libzmq C++
library (ZeroMQ) under the Mozilla Public License v2 (MPLv2).

A portion of the commits made by the Github handle "{{github username}}", with
commit author "{{github commit author}}", are copyright of {{ name }} .
This document hereby grants the libzmq project team to relicense libzmq, 
including all past, present and future contributions of the author listed above.

{{ Full Name }}  
{{ creation date of document (format: yyyy/mm/dd) }}
```

**Comment by shripchenko** [comment]
# Permission to Relicense under MPLv2

This is a statement by Sergey KHripchenko aka @shripchenko
that grants permission to relicense its copyrights in the libzmq C++
library (ZeroMQ) under the Mozilla Public License v2 (MPLv2).

A portion of the commits made by the Github handle "shripchenko", with
commit author "Sergey KHripchenko <shripchenko@intermedia.net>", are copyright of Sergey KHripchenko.
This document hereby grants the libzmq project team to relicense libzmq, 
including all past, present and future contributions of the author listed above.

Sergey KHripchenko
2023/06/07

**Comment by rlenferink** [comment]
> @rlenferink could you please transfer ownership of the https://github.com/rlenferink/libzmq-relicense repository to me? I will then transfer it to the org here, and archive it

@bluca I just transferred the repository to you for archival

**Comment by bluca** [comment]
great, thank you, will fix a rebase mistake, add the license grant from above, move it to this org and archive it
